### ISSUE TITLE
CKA_PUBLIC_KEY_INFO getting TypeInvalid

### ISSUE THREAD

**Issue by hcldan** [parallaxsecond/rust-cryptoki#142]
I hope you can give me a pointer.   I'm using a smartcard that has some certs on it for smime.

I can use the ActivClient smartcard tool on windows to poke around at the details, and I've been using this library fairly successfully pointing to their driver so far... until this.

I can see when I inspect the certs in the ActivClient program that they have a SubjectKeyIdentifier
![image](https://github.com/parallaxsecond/rust-cryptoki/assets/648319/7eb85699-6a4a-4ab3-bd0c-39c5914ac85f)

When I try to get the attribute from any of the objects returned, it always says TypeInvalid...
Is this a driver problem?  Am I misunderstanding something here?

**Comment by hcldan** [comment]
hmm... the dll says it's for pkcs11 2.20... i guess that attribute isn't in that spec.

**Comment by wiktor-k** [comment]
I'm not sure if you can query these attributes directly. IIUC you need to first get the X.509 certificate object from the card (for example using this object class: https://docs.rs/cryptoki/latest/cryptoki/object/struct.ObjectClass.html#associatedconstant.CERTIFICATE) and then parse the X.509 cert to get the field you want (the last part is outside of this library's scope).

Does it answer your question or am I completely wrong? :sweat_smile:

**Comment by hcldan** [comment]
@wiktor-k That's interesting.   I am going to try to go that route, as my options are rather limited.

FWIW, it looks like in pkcs11 2.40, those attributes are defined and *should* work.  But I found the driver I have only supports pkcs11 2.20 (which does not define those attributes)

**Comment by hcldan** [comment]
@wiktor-k I have the object_handle for the cert... how do I get the bytes?

**Comment by wiktor-k** [comment]
I think you used [`Attribute::Value`](https://docs.rs/cryptoki/latest/cryptoki/object/enum.Attribute.html#variant.Value) right? (In general it's good to paste your solution for the next person that has the same problem :sweat_smile: )

**Comment by hcldan** [comment]
@wiktor-k Yes, I was actually hoping to wrap things up with my task and contribute a few examples.

Is that something that would be welcome?   Is there a good place to put them?

**Comment by hcldan** [comment]
@wiktor-k Sent a pr.  I'm not sure the doctests will work well seeing as you need a library to init the card.

**Comment by hug-dev** [comment]
I assume that your questions were answered and I see that the PR is merged so closing for now. Feel free to re-open if there remains issues :)
